src/aspire/classification/align2d.py

```diff
@@ -1,66 +1,188 @@
 import logging
+from abc import ABC, abstractmethod
 from itertools import product
 
+import matplotlib.pyplot as plt
 import numpy as np
-from tqdm import trange
+from skimage.filters import difference_of_gaussians, window
+
+# import skimage.io
+from skimage.transform import rotate, warp_polar
+from tqdm import tqdm, trange
+
+from aspire.image import Image
+from aspire.source import ArrayImageSource
+from aspire.utils.coor_trans import grid_2d
 
 logger = logging.getLogger(__name__)
 
 
-class Align2D:
+class Align2D(ABC):
     """
     Base class for 2D Image Alignment methods.
     """
 
-    def __init__(self, basis, dtype):
+    def __init__(
+        self, alignment_basis, source, composite_basis=None, batch_size=512, dtype=None
+    ):
         """
-        :param basis: Basis to be used for any methods during alignment.
+        :param alignment_basis: Basis to be used during alignment (eg FSPCA)
+        :param source: Source of original images.
+        :param composite_basis:  Basis to be used during class average composition (eg hi res Cartesian/FFB2D)
         :param dtype: Numpy dtype to be used during alignment.
         """
 
-        self.basis = basis
+        self.alignment_basis = alignment_basis
+        # if composite_basis is None, use alignment_basis
+        self.composite_basis = composite_basis or self.alignment_basis
+        self.src = source
+        self.batch_size = batch_size
         if dtype is None:
-            self.dtype = self.basis.dtype
+            if self.composite_basis:
+                self.dtype = self.composite_basis.dtype
+            elif self.src:
+                self.dtype = self.src.dtype
+            else:
+                raise RuntimeError("You must supply a basis/src/dtype.")
         else:
             self.dtype = np.dtype(dtype)
-            if self.dtype != self.basis.dtype:
-                logger.warning(
-                    f"Align2D basis.dtype {self.basis.dtype} does not match self.dtype {self.dtype}."
-                )
 
+        if self.src and self.dtype != self.src.dtype:
+            logger.warning(
+                f"{self.__class__.__name__} dtype {dtype}"
+                "does not match dtype of source {self.src.dtype}."
+            )
+        if self.composite_basis and self.dtype != self.composite_basis.dtype:
+            logger.warning(
+                f"{self.__class__.__name__} dtype {dtype}"
+                "does not match dtype of basis {self.composite_basis.dtype}."
+            )
+
+    @abstractmethod
     def align(self, classes, reflections, basis_coefficients):
```

Comment:
We have two classes, perhaps not named as you suggest, but one that averages and another that doesn't...

I think there is a method that returns the images, but I intentionally named it differently because the input is not consistent with images used elsewhere.

I'll see if I can do this quickly, but I think you are opening a can of worms here. Will see.

Reply:
Ok. If not, I'll try my hand at it.

Reply:
I did the initial refactor (inverting the aligner/averager relationship).  It will probably take me the rest of the day (or maybe two days) to test it with the examples etc.  I will try to expand the unit tests to cover the new RC methods while the longer running example runs are happening.

Reply:
Ok sounds good.

Reply:
Locally the Sim and unit tests are okay.  Trying with the exp data, otherwise I think this round 2 review is just held up by the finufft thing.